Implement Software Update Rebuild Code

ThePitter opened the pull request:
Creates new Software Update page in Admin and enables administrators to chose to begin automated builds.
See #1249 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added Software Update admin interface with multi-language support (English, Spanish, French).
  * New admin dashboard for managing product rebuilds with real-time status tracking and progress visibility.
  * Rebuild workflows now capture and display initiator information and optional comments.

* **Enhancements**
  * Updated data display components to support numeric values.
  * Added real-time update notifications for software rebuild events.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ThePitter committed:
Initial commit on software update rebuild page. The general code is there, need to track why I am getting an internal error when trying to visit the page.

ThePitter committed:
Updated navigation with page-specific translations and implemented logic to call doProductAction. Unable to determine where to store the comment made by user however.

ThePitter committed:
Fixed formatting issues with lint.

ThePitter commented:
Everything is setup but am uncertain where the comment is supposed to be stored?

FyreByrd commented on src/routes/(authenticated)/admin/settings/software-update/+page.svelte:
```diff
@@ -0,0 +1,52 @@
+    <span class="validator-hint">{m.admin_software_update_comment_required()}</span>
+  </LabeledFormInput>
+  <input type="submit" class="btn btn-primary" value={m.admin_software_update_rebuild_start()} disabled={!propsOk}/>
+    
```
It would probably be helpful to show a list of Projects/Products that would be affected.

FyreByrd reviewed (changes requested):
As best as I can tell, all of the CodeRabbit stuff this time around is accurate; please address those.

Moving forward, please make sure you are running `npm run check` and `npm run format` and fix all reported errors before pushing.

There are some potential changes for the software updates page that @chrisvire and I had discussed briefly earlier last week; we will need to discuss those more in depth and update the issues accordingly. One of us will let you know when that happens.

ThePitter committed:
Fixed coderabbit suggested changes, formatting, and completed TODO. Added an optional comment parameter to doProductAction (and to a sub function Workflow.create) which is implemented for both starting a workflow and ending a workflow.

ThePitter committed:
Moved software update page to main menu, and added initial logic to show what organizations would be affected by the button.

ThePitter committed:
Improved readability of update page. Updated logic to properly identify what rebuilds to call. Added initial logic to begin determinig whether to show start or pause on builds, but need more informaiton on what transitions look like in the database to make a determination.

ThePitter committed:
format fixes

FyreByrd reviewed (changes requested):
This is starting to look better.
A few things for now:
- There is some duplicated/incorrect/unnecessary logic when getting a list of organizations on the server side. Please take a look at `src/routes/users/org/[[id=idNumber]]/+page.server.ts` for inspiration.
- We recently added a `SystemVersions` table with information on AppBuilder versions. Please use this to further filter projects.
- Both the load function and the action should have the same filtration logic for orgs/projects/products. Please break this out into one or more functions (in the same file) for consistency/reusability.

I will be working on creating a database table to record software updates for you to write to/read from for this PR. I will ping one of you when that's ready.

ThePitter commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,166 @@
+    // Verify user is admin of organizations selected.
+    orgs.forEach((org) => {
+      locals.security.requireAdminOfOrg(org.OrganizationId);
+      if (!searchOrgs.includes(org.OrganizationId)) searchOrgs.push(org.OrganizationId); // Make sure entries are only added once
```
I'll remove the check here then, I think I was originally just assuming it's better to have the extra verification. I did originally use `requireAdminOfOrgIn` but since it just puts the same `requireAdminOfOrg` function in a loop as well, I thought having it in one loop would be better 😅

ThePitter commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,166 @@
+    where: {
+      OrganizationId: searchOrgs ? { in: searchOrgs } : undefined,
+      DateArchived: null,
+      RebuildOnSoftwareUpdate: true
```
I think this is already happening at line 68, unless you are referencing a different way?

FyreByrd replied:
I'm realizing this may be rather complicated....

So, for what is:
`ProductBuild.AppBuilderVersion`
`Project.ApplicationTypeId`
`Organization.BuildEngineUrl` (or env.DEFAULT_BUILD_ENGINE_URL if `Organization.UseDefaultBuildEngine`)

For what should be:
`SystemVersions.*`

If what is does not match what should be, it is out of date.
Hope this helps?

ThePitter committed:
Updated localization for software update rebuild page. Added logic to only affect projects not currently up to date with with SoftwareVersion. Rewrote some code for simplicity and readability.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,179 @@
+export const load = (async ({ url, locals, params }) => {
+  // Determine what organizations are being affected
+  const searchOrgs = await getOrganizations(locals, params);
+  if (isNaN(Number(params.orgId))) locals.security.requireAdminOfOrgIn(searchOrgs);
```
You don't need to check if `param.orgId` is NaN.

FyreByrd replied:
You still need the security checks. All I'm saying is that it's not necessary to check if the organization id does or does not parse to a valid number. You just need to check if it is present.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,179 @@
+ */
+async function determineTargetOrgs(locals: App.Locals): Promise<number[]> {
+  const orgs = locals.security.isSuperAdmin
+    ? await DatabaseReads.userRoles.findMany({
```
While this technically works, it would probably be better to query organizations directly.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,179 @@
+    : await DatabaseReads.userRoles.findMany({
+        where: {
+          UserId: locals.security.userId,
+          RoleId: { in: [RoleId.SuperAdmin, RoleId.OrgAdmin] } // Must be a super admin or organization admin
```
You're already checking for Super Admin above, so you shouldn't be filtering for that here.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,179 @@
+        },
+        select: { OrganizationId: true }
+      });
+  const searchOrgs = new Set<number>();
```
I appreciate the attention to detail, but ensuring uniqueness really isn't necessary, especially if you follow my above suggestions.

FyreByrd replied:
If you want to ensure uniqueness for display purposes, you *can* do this on the client side instead? I would also recommend just passing the array of ids directly into the set constructor.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,179 @@
+            ...(currentAppBuilderVersion
+              ? {
+                  NOT: [
+                    { VersionBuilt: currentAppBuilderVersion },
```
VersionBuilt is not what you are wanting to query on. This is the version of the app, not the version of AppBuilders used to build it.

ThePitter replied:
Somehow missed that you'd replied to some of my earlier comments on where I should pull versioning from, fixing now.

FyreByrd reviewed (changes requested):
See comments below for my feedback on your most recent changes.

I am looking forward to seeing what y'all come up with for UI. If you need ideas I can send a DM of a rough sketch I had in mind.

ThePitter committed:
Rewrote the projects query to instead pull productIDs for affected productes for whatever organizations are currently being shown. Now to find products it will: get a list of products where: the product.project.organizationID is included in a list called searchOrgs, the project has not been archived, the project's setting for rebuildOnSoftwareUpdate is true, and the product's most recent productBuild.appbuilderversion is not what SystemVersions.version is when SystemVersion.BuildEngineURL equals the product.project.organization.buildengineurl and SystemVersion.ApplicationTypeId equals the project's applicationTypeId.

becca-perk committed:
Added UI enhacements

becca-perk committed:
lint fix

becca-perk commented:
I added some enhancements to the UI that were suggested here: https://github.com/sillsdev/appbuilder-portal/issues/1249#issuecomment-3514246503

However, I am still having trouble testing it because we are still not able to publish the build, and I believe that is required for the rebuilds.

becca-perk committed:
fixed loosely typed lint error

becca-perk committed:
addressed old comments and made UI enhancements for software update rebuild feature

becca-perk committed:
fixed translation

becca-perk committed:
Batch-fetch the systemVeresions

becca-perk committed:
removed debugging

becca-perk commented:
@FyreByrd, we have made some recent changes, but are still struggling with testing due to not being able to publish

becca-perk committed:
Updated to use the correct table and more UI updates

becca-perk committed:
lint fix

becca-perk committed:
Addressed hardcoded english

becca-perk commented:
To test, I kind of just fudged the AppBuilder version and deleted the workflow instance. Probably not the best, but it seemed to do the job. 
<img width="975" height="515" alt="image" src="https://github.com/user-attachments/assets/114898a4-d10f-4ad1-a2b8-c302494827fe" />
 
<img width="975" height="507" alt="image" src="https://github.com/user-attachments/assets/5b7dfcb9-12de-4faf-85b6-793c3ba6ea65" />

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,273 @@
+ * @param locals The request locals containing security information.
+ * @returns An array of organization IDs.
+ */
+async function determineTargetOrgs(locals: App.Locals): Promise<number[]> {
```
This and `getOrganizations` should be replaced by the helper function I added in [1428](https://github.com/sillsdev/appbuilder-portal/pull/1428). My helper function does NOT return a list of organization IDs; it returns a filter for a Prisma query, so further adjustments will need to be made here in this file to use said helper function.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.svelte:
```diff
@@ -0,0 +1,224 @@
+            {completedCount} / {summary.productCount}
+          </p>
+          <div
+            style="width: 100%; margin-top: 0.5rem; border-style: solid; border-color: white; border-width: 1px; border-radius: 4px; overflow: hidden; height: 24px;"
```
Why are there inline styles here? Most, if not all, of these can be replaced by TailwindCSS or DaisyUI equivalents

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.svelte:
This page should also show a list of ALL relevant software updates, not just the most recent/active one

FyreByrd reviewed (changes requested):
This is a good step forward. The main thing that is definitely missing is the ability to select which AppBuilders are being updated (i.e. SAB, DAB, etc.). 
I also have a few nitpicks and ended up writing a helper function for one of them. 

You will need to rebase on upstream develop before making further changes. 

I will need to talk with Chris about whether your polling approach on the software updates page is what we want or not.

There may be some more changes that I will eventually want made, but will either need to be discussed or I didn't want them lost in the noise of all my other feedback, so I have left them out for now.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.svelte:
```diff
@@ -0,0 +1,224 @@
+  let { data }: Props = $props();
+
+  const activeUpdateIds = persistedSession<number[]>('software-update-active-ids', []);
+  const persistedSummary = persistedSession<Summary | null>('software-update-summary', null);
```
If the user navigates away or refreshes, the list of active updates should be able to be returned by the load function if they come back to this page, right?

becca-perk committed:
Updated to use BuildEngine helpers, fixed SuperForms data handling, and removed persistent values

becca-perk commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.server.ts:
```diff
@@ -0,0 +1,273 @@
+
+const formSchema = v.object({
+  comment: v.pipe(v.string(), v.minLength(1, 'Comment is required'))
+  // Since we are only getting a comment, I do not believe we need a properties: propertiesSchema here.
```
This was Peter's comment. Not sure what he meant

judah-sotomayor committed:
fix: Replace custom helpers with filterAdminOrgs.

Also fix: Authentication needs to handle multiple orgs properly.

becca-perk committed:
Enhance software update functionality with application type filtering, improved UI, and SSE support for status updates

judah-sotomayor committed:
fix: Make the software-update page switch orgs.

becca-perk committed:
implement RebuildCard component for software updates

becca-perk committed:
feat: add support for software updates and rebuild tracking

- Extend JobType enum to include SvelteSSE_UpdateSoftwareUpdates.
- Introduce UpdateSoftwareUpdates interface in SvelteProjectSSE namespace.
- Modify recordRebuilds function to handle organization-specific rebuild requests.
- Implement getRebuildsForOrgIds function to fetch rebuilds based on organization IDs.
- Update SSE endpoint to emit rebuild updates for relevant organizations.
- Enhance RebuildCard and ApplicationTypesSelector components for better display and handling of software updates.
- Refactor code for improved readability and maintainability.

becca-perk committed:
refactor: update URL fetching method in getProductsForRebuild function

becca-perk committed:
fix: correct Spanish localization for organization title and improve error handling in SSE updates

becca-perk committed:
testing removal

becca-perk committed:
Lint fix

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/sse/+server.ts:
```diff
@@ -0,0 +1,78 @@
+
+  // Check user permissions for each organization
+  for (const orgId of orgIds) {
+    request.locals.security.requireMemberOfOrgOrSuperAdmin(orgId);
```
We have `requireAdminOfOrgIn` that already does the same thing.

FyreByrd commented on src/routes/(authenticated)/software-update/RebuildCard.svelte:
```diff
@@ -0,0 +1,101 @@
+          <span class="font-semibold mr-1">
+            {m.admin_software_update_application_type_title()}:
+          </span>
+          {rebuild.ApplicationType.Name ?? ''}
```
This should be `Description`, not `Name`

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.svelte:
```diff
@@ -0,0 +1,212 @@
+  });
+
+  // Filter products based on selected application types
+  const filteredProducts = $derived(() => {
```
All of these `filteredX` should be using `$derived.by`. You also really don't need the two `filteredXCount` ones

FyreByrd reviewed (changes requested):
This current iteration has a lot of unnecessarily duplicated data. You should look for ways to simplify things directly in the database queries themselves as much as possible, rather than relying on post query filtering. It may also be helpful to use Projects (and/or Organizations) as the base data model for most of your queries here, rather than Products. This should hopefully simplify a lot of your helper functions as well (and maybe remove the need for some of them...).

FyreByrd commented on src/routes/(authenticated)/software-update/RebuildCard.svelte:
Yeah. That looks good to me. If I end up needing to change something later on, I can always do that myself later. I do have a pretty big PR in the works to add more icons.

becca-perk committed:
refactor: streamline data handling in software updates, improve error logging, and edit style of rebuildcard

becca-perk committed:
lint fixes

judah-sotomayor committed:
refactor(sse): Use body instead of query parameter

Instead of passing organization IDs in a query-string parameter, use the
post body.
This avoids parsing.

judah-sotomayor committed:
refactor: Clean up getProductsForRebuild.

This generally reduces the entire getProductsForRebuild to two
query+map/reduce operations, one to fetch the current version table and
the other to fetch and filter through products.

judah-sotomayor committed:
Refactor prisma queries into a single query

judah-sotomayor committed:
Fix: Filter out projects with no products.

judah-sotomayor committed:
Fix issues caused by rebase.

judah-sotomayor committed:
Complete refactor of load function.

Once again overhauling load to further simplify the query and data
returned by it.

judah-sotomayor committed:
Lint fix.

judah-sotomayor committed:
Simplify the software updates page.

This temporarily removes SSE functionality for the sake of making sure
MVP is there.

Basically, provide a stable framework to build on top of to provide SSE.

Queries are much simplified. The function to get rebuilds has been
redesigned to only fetch rebuilds from either one organization or every
organization the user is a member of. This streamlines the code.
I have also removed a few files and consolidated their code into other
files: sse.ts into the software-updates/index.ts file, and
ApplicationTypesSelector into a built-in section of the main
page.svelte.

judah-sotomayor committed:
Reintroduce SSE

Add SSE back, adjusted to take advantage of the simplified software
update functions.

judah-sotomayor committed:
Correct prisma migration.

FyreByrd commented on src/lib/prisma/migrations/20260414013742_37_simplify_software_updates/migration.sql:
Please remove the timestamp from the filename.

judah-sotomayor replied:
Done.

FyreByrd commented on src/lib/server/database/SoftwareUpdates.ts:
```diff
@@ -0,0 +1,103 @@
+  organizationId: number;
+};
+
+export async function create(data: RequirePrimitive<Prisma.SoftwareUpdatesUncheckedCreateInput>) {
```
Why is this using a transaction? Is this needed?

judah-sotomayor replied:
Because I copied code from elsewhere in the codebase and it worked, so I didn't investigate too closely. I can swap for a normal create?

FyreByrd replied:
It probably doesn't matter too much? Transaction may have a blocking connection. It's fine as is, but not super necessary. We really don't need to be using transactions unless it's truly necessary to prevent race conditions.

judah-sotomayor replied:
Fixed.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/sse/+server.ts:
```diff
@@ -0,0 +1,58 @@
+      try {
+        const overlap = orgIds.filter((u) => organizations.includes(u));
+
+        if (overlap) {
```
This won't work like you expect. Unfortunately, JS considers an empty array to be truthy.

judah-sotomayor replied:
Fixed.

FyreByrd commented on src/routes/(authenticated)/admin/settings/+layout.svelte:
```diff
@@ -23,6 +23,8 @@
     },
     { text: m.admin_nav_prodDefs(), route: 'product-definitions', icon: Icons.Product },
+    // TODO Get icon for software update
+    { text: m.admin_nav_software_update(), route: 'software-update', icon: Icons.Product },
```
This shouldn't go in the admin settings layout. Just the root layout sidebar. Make sure it's hidden for non-admins.

judah-sotomayor replied:
Fixed.

FyreByrd commented on src/lib/server/database/SoftwareUpdates.ts:
```diff
@@ -0,0 +1,103 @@
+/**
+ * Helper: Get all unique project IDs that contain the given products
+ */
+async function getProjectIdsFromProductIds(productIds: string[]): Promise<number[]> {
```
You're only using this function in the one location. If you're using it in one place, non-recursively, you shouldn't need to have it extracted out into a separate helper function. If you instead have all of this where it's used, you should be able to come up with a much more efficient query that still fulfills what you need. Based on the one place you use it, you could probably replace it with a `count` database call.

judah-sotomayor replied:
I'm actually not certain what the current use of it is for. I'm removing it entirely unless @becca-perk has any objection

becca-perk replied:
You're right, it does not need to be a helper function. @judah-sotomayor I think put it in as a guard to avoid sending an SSE queue job when there is nothing meaningful to notify. The helper could be removed, but if we still want to have a guard here, we would need to work on simplifying it

FyreByrd replied:
That should be correct? Since you're doing the SSE on orgIds?

FyreByrd commented on src/lib/server/database/SoftwareUpdates.ts:
```diff
@@ -0,0 +1,103 @@
+    for (const p of u.Products) {
+      // Require a successful build at the target version at or after the update start time
+      // TODO What is the correct procedure when the update does not have a date created?
+      if (p.ProductBuilds[0].DateCreated < u.DateCreated) {
```
ProductBuilds.DateCreated is still possibly null. Also, using `valueOf` would be recommended here for ordering comparison.

judah-sotomayor replied:
Fixed.

FyreByrd commented on src/lib/server/database/SoftwareUpdates.ts:
```diff
@@ -0,0 +1,103 @@
+ * Checks if a specific product's successful build completes any open SoftwareUpdates.
+ * Called after each successful build for immediate completion detection.
+ */
+export async function completeForProduct(productId: string): Promise<void> {
```
So, if I understand this correctly, an update is marked as complete as soon as any one product within it has been built successfully?

FyreByrd replied:
Sorry to complicate things, but we may need a bit more metadata on the relation between products and software updates...

judah-sotomayor replied:
Will be resolved as per our call and discord discussion.

FyreByrd commented on src/lib/server/database/SoftwareUpdates.ts:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    if (ok) {
+      const orgIds = [...new Set<number>(u.Products.map((p) => p.Project.Organization.Id))];
```
If this function is just for marking a single product as having completed its portion of the update, you only need to update SSE for one org.

judah-sotomayor replied:
It is meant to run on the end of every product build in order to check if this product is the last one a software update is waiting for. If it is, close the software update.

FyreByrd commented on src/lib/software-updates/components/RebuildCard.svelte:
```diff
@@ -0,0 +1,83 @@
+          {rebuild.InitiatedBy ?? ''}
+        </span>
+        <span class="flex items-center mb-1">
+          <IconContainer icon="mdi:package-variant" width={20} class="mr-1 shrink-0" />
```
This icon and `mdi:folder-multiple` will either need to be replaced with something else or be added to the `Icons` object in `$lib/icons`.

judah-sotomayor replied:
Fixed

FyreByrd commented on src/lib/software-updates/index.ts:
This file has server-side functionality in it. Please split it into separate index and server files.

judah-sotomayor replied:
Done.

FyreByrd reviewed (changes requested):
This looks good so far. I think we will need to talk a bit about what metadata we need to keep track of, as well as a good strategy for cancelling updates.

becca-perk committed:
Add error handling for form validation in software updates page

judah-sotomayor committed:
feat: Add Aiden's changes to the database.

- Consolidate software update migrations.
- Break out relation between Products and SoftwareUpdates into explicit
  m-n table.
- Add WorkflowInstances to SoftwareUpdates to support mass cancellation.

judah-sotomayor committed:
fix: Remove tx from SoftwareUpdates create.

judah-sotomayor committed:
fix: Check array length rather than truthyness.

FyreByrd commented on src/routes/(authenticated)/software-update/[[orgId=idNumber]]/+page.svelte:
```diff
@@ -156,11 +156,13 @@
+        {#if $errors.comment}
+          <span class="text-error text-sm">{m.admin_software_update_comment_required()}</span>
+        {/if}
+        <!--<span class="validator-hint">{m.admin_software_update_comment_required()}</span>-->
```
@becca-perk You didn't need to make this change. This should have been fine as is and would have matched the error handling we have elsewhere.

judah-sotomayor replied:
There was an issue going on where after submitting the form for a software update the comment box would pop up a validation error saying you need a comment to submit the form. This commit was supposed to fix that, but if it's not the right approach do you have any suggestions?

FyreByrd replied:
Do you want the form to reset after it's been submitted? If so, that may be where the error is coming from... let me look into it real quick.

FyreByrd replied:
Ok. If you want to do it this way, it's probably fine. Otherwise I found one of the few locations that has a similar problem. You can look at `/projects/[id]/forms/Reviewers.svelte` if you want to see how I went about resolving it.

judah-sotomayor committed:
fix: Remove getProjectIdsFromProductIds().

judah-sotomayor committed:
fix: Null-check ProductBuilds.DateCreated.

judah-sotomayor committed:
fix: Remove softwareupdates entry from admin layout.

judah-sotomayor committed:
fix: Add icons on RebuildCard and sidebar.

Add icons to RebuildCard using the `Icons` construct.
Also ensure the software updates button has an icon in the sidebar.

judah-sotomayor committed:
fix: Split out server code from software-updates/index

judah-sotomayor committed:
fix: Check orgIds length before notifying sses.

judah-sotomayor committed:
fix: Lint fixes

judah-sotomayor committed:
fix: Remove implicit m-n between SoftwareUpdates and Products.

judah-sotomayor committed:
fix: Correct product filter toggles.

judah-sotomayor committed:
feat: Finalize database migration.

judah-sotomayor committed:
feat: Adapt software updates to latest database schema.

Ensure connection between workflows and software updates table.
Ensure connection between products and software updates.

judah-sotomayor committed:
Preliminary code for cancellations
